gremlin/ui/device.py

```diff
@@ -30,13 +32,164 @@
+            the input are not defined, input_name() returns a default
+            input name, e.g. Axis 1 or Button 1 or Hat 1.
+
+            Note that some devices have configurable number of buttons and/or axes
+            so adjusting device database information might be required in the future.
+        """
+        input_name_display_mode = Configuration().value(
+            "global", "input-names", "input-name-display-mode"
+            )
+
+        if (
+            input_name_display_mode == "Numerical" or
+            input_name not in self._input_map
+        ):
+            return input_name
+
+        db_input_name = self._input_map[input_name].strip()
+
+        # return default input name if input is defined, but empty
+        if len(db_input_name) == 0:
+            return input_name
+
+        # return configured input name style
+        if input_name_display_mode == "Numerical + Label":
+            return f"{input_name} - {db_input_name}"
+        if input_name_display_mode == "Label":
+            return db_input_name
+
+
+
+@SingletonDecorator
+class DeviceDatabase:
+    """Provides button/axis/hat to name mapping for known devices"""
+
+    def __init__(self) -> None:
+        self._load()
+
+    def _load(self) -> None:
+
+        db_file = util.resource_path("device_db.json")
+        if not util.file_exists_and_is_accessible(db_file):
+            return
+
+        load_successful = False
+        json_data = {}
+        try:
+            json_data = json.load(open(db_file))
+            load_successful = True
+        except JSONDecodeError as error:
+            logging.getLogger("system").error(
+                    f"There was an error loading {db_file}: {error}")
+
+        parser_successful = self._parse_device_db(json_data)
+
+        if load_successful and parser_successful:
+            self._device_db = json_data
+        else:
+            self._device_db = {"revision": 0, "devices": [], "mapping": {}}
+
+    def _parse_device_db(self, device_db: Dict[str, Any]) -> bool:
+        """
+        Parse device_db to make sure it has a valid structure.
+
+        This piece could be refactored with JSON schema validation code,
+        if need be in the future.
+        """
+        parser_successful = True
+
+        # check top structure: devices and mappings
+        if (
+            device_db.get("mapping", None) is None or
+            device_db.get("devices", None) is None or
+            not isinstance(device_db.get("mapping", []), dict) or
+            not isinstance(device_db.get("devices", []), list)
+        ):
+            logging.getLogger("system").error(
+                    "DeviceDatabase is corrupt and/or is missing mapping or device information.")
+            parser_successful = False
+
+        # check devices contain mandatory fields
+        device_count = 0
+        for dev in device_db.get("devices", []):
+            if (
+                dev.get("product_id", None) is None or
```

Comment:
Technically this could be made a bit more concise and possibly clearer with this to not have to rely on getting `None` via default value.
`if not("product_id" in dev and "vendor_id" in dev and "mapping" in dev):`